src/undocommand.js

```diff
@@ -0,0 +1,200 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+import Command from '../command/command.js';
+
+/**
+ * Undo command stores batches in itself and is able to and apply reverted versions of them on the document.
+ *
+ * undo.UndoCommand
+ */
+export default class UndoCommand extends Command {
+	/**
+	 * @see engine.command.Command
+	 * @param {engine.Editor} editor
+	 */
+	constructor( editor ) {
+		super( editor );
+
+		/**
+		 * Batches which are saved by the command. They can be reversed.
+		 *
+		 * @private
+		 * @member {Array.<engine.treeModel.Batch>} undo.UndoCommand#_batchStack
+		 */
+		this._batchStack = [];
+
+		/**
+		 * For each batch, it stores the information about selection needed to recreate it after undo.
+		 *
+		 * @private
+		 * @member {WeakMap} undo.UndoCommand#_batchSelection
+		 */
+		this._batchSelection = new WeakMap();
+	}
+
+	/**
+	 * Stores a batch in the command. Stored batches can be then reverted.
+	 *
+	 * @param {engine.treeModel.Batch} batch Batch to add.
+	 * @param {Array.<engine.treeModel.Range>} selectionRanges All ranges that were in the selection when batch got
+	 * created.
+	 */
+	addBatch( batch ) {
+		this._batchStack.push( batch );
+		this._batchSelection.set(
+			batch,
+			{
+				ranges: Array.from( this.editor.document.selection.getRanges() ),
+				isBackward: this.editor.document.selection.isBackward
+			}
+		);
+	}
+
+	/**
+	 * Removes all batches from the stack.
+	 */
+	clearStack() {
+		this._batchStack = [];
+	}
+
+	/**
+	 * Checks whether this command should be enabled. Command is enabled when it has any batches in its stack.
+	 *
+	 * @private
+	 * @returns {Boolean}
+	 */
+	_checkEnabled() {
+		return this._batchStack.length > 0;
+	}
+
+	/**
+	 * Executes the command: reverts a {@link engine.treeModel.Batch batch} added to the command's stack,
+	 * applies it on the document and removes the batch from the stack.
+	 *
+	 * Fires `undo` event with reverted batch as a parameter.
+	 *
+	 * @private
```

Comment:
I belive this is a mistake in `Command` class. `_doExecute` should be `private` and should be fired only by `_execute`. Take a look at `AttributeCommand`. If you agree that this is a mistake, I'll change it in `ckeditor5` repo.